File: js/game_of_life.js
var GameOfLife = function(matrix_size){
  var _this = this;
  _this.matrix_size = matrix_size;
  _this.matrix = new Matrix(matrix_size);
  _this.cells = new Array();

  _this.generate_cells = function(){
    _.times(_this.matrix.size, function(y){
      var element = new Array(_this.matrix.size);
      _.times(_this.matrix.size, function(x){
        var cell = new Cell(x, y);
        _this.cells.push(cell);
      });
    });
  }
  _this.next_tick = function(){
    _.each(_this.cells, function(cell){
      if (cell.cache_alive && !cell.should_stay_live(_this.cells)) { cell.die(); };
      if (cell.cache_dead() && cell.should_be_reborn(_this.cells)) { cell.reborn(); };
    });
    _.each(_this.cells, function(cell){
      cell.cache_alive = cell.alive;
    });
  }
  _this.print_cells = function(){
    $("#table tbody").html("");
    _.times(_this.matrix_size, function(y){
      $("#table tbody").append("<tr id=" + "y_" + y + ">");
      _.times(_this.matrix_size, function(x){
        var cell = _this.find_cell_by_x_and_y(x, y);
        var text = cell.alive ? "O" : "-";
        $("#y_" + y).append("<td>" + text + "</td>");
      });
      $("#table tbody").append("</tr>")
    });
  }
  _this.find_cell_by_x_and_y = function(x, y){
    return _.find(_this.cells, function(cell){
      return (cell.x == x && cell.y == y);
    });
  }

  return _this;
}

var Cell = function(x, y){
  var _this = this;
  _this.x = x;
  _this.y = y;
  _this.alive = false;
  _this.cache_alive = false;

  _this.neighbours_among = function(cells){
    _this.mates = _.without(cells, _this);
    return _.filter(_this.mates, function(cell){
      return _this.neighbour(cell);
    });
  }
  _this.alive_neighbours_among = function(cells){
    return _.filter(_this.neighbours_among(cells), function(cell){
      return cell.cache_alive;
    });
  }
  _this.dead_neighbours_among = function(cells){
    return _.filter(_this.neighbours_among(cells), function(cell){
      return cell.cache_dead();
    });
  }
  _this.cache_dead = function(){
    return !_this.cache_alive;
  }
  _this.dead = function(){
    return !_this.alive;
  }
  _this.neighbour = function(cell){
    return (_this.distance_from_x(cell) < 2 && _this.distance_from_y(cell) < 2);
  }
  _this.distance_from_x = function(cell){
    return Math.abs(cell.x - x);
  }
  _this.distance_from_y = function(cell){
    return Math.abs(cell.y - y);
  }
  _this.should_stay_live = function(cells){
    // console.log(_this.alive_neighbours_among(cells).length);
    return ((_this.alive_neighbours_among(cells).length == 2 || _this.alive_neighbours_among(cells).length == 3) && _this.alive)
  }
  _this.should_be_reborn = function(cells){
    return (_this.alive_neighbours_among(cells).length == 3 && _this.dead())
  }
  _this.die = function(){
    _this.alive = false;
  }
  _this.reborn = function(){
    _this.alive = true;
  }

  return _this;
}

var Matrix = function(size){
  var _this = this;
  _this.size = size;
  // _this.structure = new Array(_this.size);
  // _.times(_this.size, function(n){
  //   _this.structure[n] = new Array(_this.size);
  // });
  return _this;
}

var game = new GameOfLife(5);
game.generate_cells();

game.cells[0].cache_alive = game.cells[0].alive = true;
game.cells[1].cache_alive = game.cells[1].alive = true;
game.cells[2].cache_alive = game.cells[2].alive = true;
game.cells[3].cache_alive = game.cells[3].alive = true;
game.cells[5].cache_alive = game.cells[5].alive = true;
game.cells[7].cache_alive = game.cells[7].alive = true;
game.cells[8].cache_alive = game.cells[8].alive = true;
game.cells[9].cache_alive = game.cells[9].alive = true;
game.cells[11].cache_alive = game.cells[11].alive = true;
game.cells[20].cache_alive = game.cells[20].alive = true;
game.cells[21].cache_alive = game.cells[21].alive = true;
game.cells[23].cache_alive = game.cells[23].alive = true;

game.print_cells();
timer = setInterval(function(){
  game.next_tick();
  game.print_cells();
}, 5000);

File: ruby/game_of_life.rb
class GameOfLife

  attr_accessor :matrix, :cells, :matrix_size

  def initialize matrix_size=10
    @matrix_size, @matrix, @cells = matrix_size, Matrix.new(matrix_size), []
  end

  def generate_cells
    matrix.structure.each_with_index do |element, y|
      element.each_with_index do |element, x|
        @cells << Cell.new(x, y)
      end
    end
  end

  def next_tick
    cells.each do |cell|
      cell.die if cell.cache_alive? && !cell.show_stay_live?(cells)
      cell.reborn if cell.cache_dead? && cell.should_be_reborn?(cells)
    end
    cells.each {|cell| cell.cache_alive = cell.alive}
  end

  def print_cells
    matrix_size.times do |y|
      matrix_size.times do |x|
        cell = find_cell_by_x_and_y(x, y)
        cell.alive? ? print("O") : print("-")
      end
      print "\n"
    end
    print "\n"
  end

  def find_cell_by_x_and_y x, y
    cells.select {|cell| cell.x == x && cell.y == y}.first
  end
end

class Cell

  attr_accessor :x, :y, :alive, :mates, :cache_alive

  def initialize x, y, alive=false
    @x, @y, @alive, @cache_alive = x, y, alive, alive
  end

  def neighbours_among cells
    @mates = cells - [self]
    mates.select do |cell|
      neighbour?(cell)
    end
  end

  def alive_neighbours_among cells
    neighbours_among(cells).select { |cell| cell.cache_alive? }
  end

  def dead_neighbours_among cells
    neighbours_among(cells).select { |cell| cell.cache_dead? }
  end

  def cache_alive?
    cache_alive
  end

  def cache_dead?
    !cache_alive
  end

  def alive?
    alive
  end

  def dead?
    !alive
  end

  def neighbour? cell
    distance_from_x(cell) < 2 && distance_from_y(cell) < 2
  end

  def distance_from_x cell
    (cell.x - x).abs
  end

  def distance_from_y cell
    (cell.y - y).abs
  end

  def show_stay_live? cells
    (alive_neighbours_among(cells).count == 2 ||
      alive_neighbours_among(cells).count == 3) && alive?
  end

  def should_be_reborn? cells
    alive_neighbours_among(cells).count == 3 && dead?
  end

  def die
    @alive = false
  end

  def reborn
    @alive = true
  end
end

class Matrix

  attr_accessor :size, :structure

  def initialize size
    @size = size
    @structure = Array.new(size) { 
      Array.new(size)
    }
  end

  def get_x_within_matrix
    rand(size)
  end

  def get_y_within_matrix
    rand(size)
  end
end
